### PR TITLE
Use MTLHeap when possible for resource allocation

### PR DESCRIPTION
**Pull request by JohnColanduoni**
Adds MTLHeap support to the Metal backend, and uses it whenever possible for memory allocations.

This isn't quite ready to merge, https://github.com/gfx-rs/metal-rs/pull/7 needs to make it into crates.io first.

### CODE REVIEW

File: src/backend/metalll/src/conversions.rs [hunk 1]
```diff
@@ -120,3 +120,62 @@ pub fn map_vertex_format(format: Format) -> Option<MTLVertexFormat> {
...
+    if properties.contains(memory::WRITE_COMBINED) {
+        options |= MTLResourceOptionCPUCacheModeWriteCombined;
+    }
+    options
+}
+
+pub fn map_heap_properties_to_storage_and_cache(properties: memory::HeapProperties) -> (MTLStorageMode, MTLCPUCacheMode) {
+    let storage;
+    let cpu;
+    if properties.contains(memory::CPU_VISIBLE) {
```
Comment (kvark) [inline src/backend/metalll/src/conversions.rs]: any reason you don't do `let cpu = if {..} else {..};` and the same for `storage`?

File: src/backend/metalll/src/factory.rs [hunk 2]
```diff
@@ -45,9 +47,24 @@ impl Drop for Factory {
...
+        MTLFeatureSet::iOS_GPUFamily3_v2,
+        MTLFeatureSet::iOS_GPUFamily1_v4,
+        MTLFeatureSet::iOS_GPUFamily2_v4,
+        MTLFeatureSet::iOS_GPUFamily3_v3,
+
+        MTLFeatureSet::tvOS_GPUFamily1_v2,
+        MTLFeatureSet::tvOS_GPUFamily1_v3,
+
+        MTLFeatureSet::macOS_GPUFamily1_v3,
+    ].iter().cloned().any(|x| device.supports_feature_set(x));
```
Comment (kvark) [inline src/backend/metalll/src/factory.rs]: nice!

File: src/backend/metalll/src/factory.rs [hunk 3]
```diff
@@ -642,43 +659,83 @@ impl CoreFactory<Resources> for Factory {
...
+            size: max_size,
+            alignment: max_alignment,
         }
     }
 
-    #[cfg(not(feature = "native_heap"))]
     fn bind_buffer_memory(&mut self, heap: &n::Heap, offset: u64, buffer: n::UnboundBuffer) -> Result<n::Buffer, buffer::CreationError> {
-        Ok(n::Buffer(buffer.0))
+        let bound_buffer;
+        match *heap {
```
Comment (kvark) [inline src/backend/metalll/src/factory.rs]: similarly, please convert to functional style aka `let bound_buffer = match *heap {...};`

**Comment by kvark** [comment]
@JohnColanduoni thanks for updating it!
Do you think we still need to merge this in the light of #1384? Or should we postpone the decision until we figure out the plan on how exactly to make `MTLHeap` fitting our API?

**Comment by JohnColanduoni** [comment]
@kvark This at least fixes the handling of memory types (previously everything was allocated in coherent, cpu visible memory) with needs to be done regardless of what we decide on #1384. Also if we decide to use heaps and fall back on driver-managed allocation when that fails we're just a few lines away.

**Pull request merged**